File: foundations-jdbc-scala/src/scala/dev/typr/foundationssc/PgType.scala
package dev.typr.foundationssc

class PgType[T](override val underlying: dev.typr.foundations.PgType[T]) extends DbType[T](underlying):
  override def opt: PgType[Option[T]] =
    PgType(underlying.opt().to(Bijections.optionalToOption))

  override def to[B](bijection: Bijection[T, B]): PgType[B] =
    PgType(underlying.to(bijection))

  def transform[B](f: T => B, g: B => T): PgType[B] =
    PgType(underlying.transform(v => f(v), v => g(v)))

  /** Variable-length PG array of this type — Scala-side as immutable {@code List[T]}. */
  def array: PgType[List[T]] = PgType(
    underlying
      .array()
      .transform(
        jlist => scala.jdk.CollectionConverters.ListHasAsScala(jlist).asScala.toList,
        slist => {
          val al = new java.util.ArrayList[T](slist.size)
          slist.foreach(al.add)
          al
        }
      )
  )

  def encode(value: T): Fragment = new Fragment(underlying.encode(value))

  def pgText(): PgText[T] = underlying.pgText()

  def withTypename(typename: PgTypename[T]): PgType[T] = PgType(underlying.withTypename(typename))
  def withTypename(sqlType: String): PgType[T] = PgType(underlying.withTypename(sqlType))
  def renamed(value: String): PgType[T] = PgType(underlying.renamed(value))
  def renamedDropPrecision(value: String): PgType[T] = PgType(underlying.renamedDropPrecision(value))

  def withRead(read: PgRead[T]): PgType[T] = PgType(underlying.withRead(read))
  def withWrite(write: PgWrite[T]): PgType[T] = PgType(underlying.withWrite(write))
  def withText(text: PgText[T]): PgType[T] = PgType(underlying.withText(text))
  def withJson(json: PgJson[T]): PgType[T] = PgType(underlying.withJson(json))
  def withOutParam(outParam: PgOutParam[T]): PgType[T] = PgType(underlying.withOutParam(outParam))
  def withArrayCodec(codec: PgElementCodec[T]): PgType[T] = PgType(underlying.withArrayCodec(codec))
  def withAnalysis(opts: AnalysisOptions): PgType[T] = PgType(underlying.withAnalysis(opts))

  def unchecked(): PgType[T] = PgType(underlying.unchecked())
  def nullableOk(): PgType[T] = PgType(underlying.nullableOk())

File: foundations-jdbc-kotlin/src/kotlin/dev/typr/foundationskt/PgType.kt
@file:Suppress("unused")
package dev.typr.foundationskt

import dev.typr.foundations.AnalysisOptions
import dev.typr.foundations.PgElementCodec
import dev.typr.foundations.PgJson
import dev.typr.foundations.PgOutParam
import dev.typr.foundations.PgRead
import dev.typr.foundations.PgText
import dev.typr.foundations.PgTypename
import dev.typr.foundations.PgWrite

class PgType<T>(override val underlying: dev.typr.foundations.PgType<T>) : DbType<T>(underlying) {
    override fun opt(): PgType<T?> =
        PgType(underlying.opt().to(Bijection.optionalToNullable()))

    override fun <B> to(bijection: Bijection<T, B>): PgType<B> =
        PgType(underlying.to(bijection.underlying))

    fun <B> transform(f: (T) -> B, g: (B) -> T): PgType<B> =
        PgType(underlying.transform(dev.typr.foundations.SqlFunction { f(it) }, g))

    /** Variable-length PG array of this type — Kotlin-side as {@code List<T>}. */
    fun array(): PgType<List<T>> = PgType(underlying.array())

    fun encode(value: T): Fragment = Fragment(underlying.encode(value))

    fun pgText(): PgText<T> = underlying.pgText()

    fun withTypename(typename: PgTypename<T>): PgType<T> = PgType(underlying.withTypename(typename))
    fun withTypename(sqlType: String): PgType<T> = PgType(underlying.withTypename(sqlType))
    fun renamed(value: String): PgType<T> = PgType(underlying.renamed(value))
    fun renamedDropPrecision(value: String): PgType<T> = PgType(underlying.renamedDropPrecision(value))

    fun withRead(read: PgRead<T>): PgType<T> = PgType(underlying.withRead(read))
    fun withWrite(write: PgWrite<T>): PgType<T> = PgType(underlying.withWrite(write))
    fun withText(text: PgText<T>): PgType<T> = PgType(underlying.withText(text))
    fun withJson(json: PgJson<T>): PgType<T> = PgType(underlying.withJson(json))
    fun withOutParam(outParam: PgOutParam<T>): PgType<T> = PgType(underlying.withOutParam(outParam))
    fun withArrayCodec(codec: PgElementCodec<T>): PgType<T> = PgType(underlying.withArrayCodec(codec))
    fun withAnalysis(opts: AnalysisOptions): PgType<T> = PgType(underlying.withAnalysis(opts))

    fun unchecked(): PgType<T> = PgType(underlying.unchecked())
    fun nullableOk(): PgType<T> = PgType(underlying.nullableOk())
}
